File: app/src/main/java/rod/bailey/trafficatnsw/MainActivity.java
package rod.bailey.trafficatnsw;

import android.app.Fragment;
import android.app.FragmentManager;
import android.content.Context;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.v4.app.ActionBarDrawerToggle;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ListView;

import rod.bailey.trafficatnsw.cameras.TrafficCameraListFragment;
import rod.bailey.trafficatnsw.hazard.HazardsFragment;
import rod.bailey.trafficatnsw.traveltime.TravelTimesFragment;
import rod.bailey.trafficatnsw.util.ConfigSingleton;
import rod.bailey.trafficatnsw.util.MLog;

import static rod.bailey.trafficatnsw.cameras.TrafficCameraListFragment.*;
import static rod.bailey.trafficatnsw.traveltime.TravelTimesFragment.*;

public class MainActivity extends AppCompatActivity {

    private static final String CONFIG_PROPERTIES_FILE_NAME = "config.properties";

    @SuppressWarnings("unused")
    private static final String LOG_TAG = MainActivity.class.getSimpleName();

    private ActionBarDrawerToggle drawerToggle;

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        drawerToggle.onConfigurationChanged(newConfig);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ConfigSingleton config = ConfigSingleton.getInstance();
        config.init(CONFIG_PROPERTIES_FILE_NAME, this);

        MLog.init(config, this);

        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        android.support.v7.app.ActionBarDrawerToggle toggle = new android.support.v7.app.ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.setDrawerListener(toggle);
        toggle.syncState();

        NavigationView navView = (NavigationView) findViewById(R.id.navigation_drawer_view);
        navView.setNavigationItemSelectedListener(new DrawerItemClickListener(drawer));

    }

    @Override
    public View onCreateView(String name, Context context, AttributeSet attrs) {
        View view = super.onCreateView(name, context, attrs);
        return view;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (drawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
//		drawerToggle.syncState();
    }

    /**
     * Listens for a tap on an item in the navigation drawer and invokes the
     * corresponding Fragment. Also changes the title approriately.
     */
    public class DrawerItemClickListener implements NavigationView.OnNavigationItemSelectedListener {

        private DrawerLayout drawerLayout;

        public DrawerItemClickListener(DrawerLayout drawerLayout) {
            this.drawerLayout = drawerLayout;
        }

        @Override
        public boolean onNavigationItemSelected(MenuItem item) {

            Log.d(LOG_TAG, "onNavigationItemSelected: item=" + item);

            switch (item.getItemId()) {
                case R.id.menu_item_sydney_incidents:
                    navigateToIncidents(HazardsFragment.ARG_MODE_VALUE_SYDNEY);
                    break;

                case R.id.menu_item_regional_nsw_incidents:
                    navigateToIncidents(HazardsFragment.ARG_MODE_VALUE_REGIONAL);
                    break;

                case R.id.menu_item_favourite_cameras:
                    navigateToTrafficCamera(TrafficCameraListFragment.ARG_MODE_VALUE_FAVOURITES);
                    break;

                case R.id.menu_item_sydney_cameras:
                    navigateToTrafficCamera(TrafficCameraListFragment.ARG_MODE_VALUE_SYDNEY);
                    break;

                case R.id.menu_item_regional_nsw_cameras:
                    navigateToTrafficCamera(TrafficCameraListFragment.ARG_MODE_VALUE_REGIONAL);
                    break;

                case R.id.menu_item_travel_times_m1:
                    navigateToTravelTimes(TravelTimesFragment.ARG_MWAY_VALUE_M1);
                    break;

                case R.id.menu_item_travel_times_m2:
                    navigateToTravelTimes(TravelTimesFragment.ARG_MWAY_VALUE_M2);
                    break;

                case R.id.menu_item_travel_times_m3:
                    navigateToTravelTimes(TravelTimesFragment.ARG_MWAY_VALUE_M4);
                    break;

                case R.id.menu_item_travel_times_m4:
                    navigateToTravelTimes(TravelTimesFragment.ARG_MWAY_VALUE_M7);
                    break;
            } // switch

            drawerLayout.closeDrawer(Gravity.LEFT);
            return true;

        } // onNavigationItemSelected()

        private void navigateToIncidents(int argHazardModeValue) {
            Fragment fragment = new HazardsFragment();

            Bundle bundle = new Bundle();
            bundle.putInt(ARG_MODE_KEY, argHazardModeValue);
            fragment.setArguments(bundle);

            // Insert the fragment by replacing any existing fragment
            FragmentManager fragmentManager = getFragmentManager();
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, fragment).commit();
        }

        private void navigateToTrafficCamera(int argCameraModeValue) {
            Fragment fragment = new TrafficCameraListFragment();

            Bundle bundle = new Bundle();
            bundle.putInt(ARG_MODE_KEY, argCameraModeValue);
            fragment.setArguments(bundle);

            // Insert the fragment by replacing any existing fragment
            FragmentManager fragmentManager = getFragmentManager();
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, fragment).commit();
        }

        private void navigateToTravelTimes(String argTravelTimeValue) {
            Fragment fragment = new TravelTimesFragment();
            Bundle bundle = new Bundle();
            bundle.putString(ARG_MWAY_KEY, argTravelTimeValue);
            fragment.setArguments(bundle);

            FragmentManager fragmentManager = getFragmentManager();
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, fragment).commit();
        }
    } // class DrawItemClickListener

}

File: app/src/main/java/rod/bailey/trafficatnsw/MainActivity.kt
package rod.bailey.trafficatnsw

import android.content.res.Configuration
import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.Gravity
import android.view.MenuItem
import org.androidannotations.annotations.AfterViews
import org.androidannotations.annotations.EActivity
import rod.bailey.trafficatnsw.app.TrafficAtNSWApplication
import rod.bailey.trafficatnsw.cameras.TrafficCameraListFragment
import rod.bailey.trafficatnsw.hazard.HazardListFragment
import rod.bailey.trafficatnsw.hazard.HazardListMode
import rod.bailey.trafficatnsw.traveltime.TravelTimesFragment
import rod.bailey.trafficatnsw.traveltime.config.Motorway
import rod.bailey.trafficatnsw.util.ConfigSingleton
import rod.bailey.trafficatnsw.util.MLog
import javax.inject.Inject

@EActivity(R.layout.activity_main)
open class MainActivity : AppCompatActivity() {

	init {
		TrafficAtNSWApplication.graph.inject(this)
	}

	private lateinit var drawerToggle: ActionBarDrawerToggle

	@Inject
	lateinit var config: ConfigSingleton

	override fun onConfigurationChanged(newConfig: Configuration) {
		super.onConfigurationChanged(newConfig)
		drawerToggle.onConfigurationChanged(newConfig)
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		// Initialize config facility
		config.init(CONFIG_PROPERTIES_FILE_NAME, this)

		// This logging facility needs initialization
		MLog.init(config)
	}

	@AfterViews
	fun afterViews() {
		val toolbar = findViewById(R.id.toolbar) as Toolbar
		setSupportActionBar(toolbar)

		// Link drawer toggle icon to nav drawer open/close state
		val drawer = findViewById(R.id.drawer_layout) as DrawerLayout
		drawerToggle = ActionBarDrawerToggle(
			this,
			drawer,
			toolbar,
			R.string.app_navigation_drawer_open,
			R.string.app_navigation_drawer_close)
		drawer.addDrawerListener(drawerToggle)
		drawerToggle.syncState()

		// Navigation
		val navView = findViewById(R.id.navigation_drawer_view) as NavigationView
		navView.setNavigationItemSelectedListener(DrawerItemClickListener(drawer))
	}

	override fun onOptionsItemSelected(item: MenuItem): Boolean {
		if (drawerToggle.onOptionsItemSelected(item)) {
			return true
		}
		return super.onOptionsItemSelected(item)
	}

	override fun onPostCreate(savedInstanceState: Bundle?) {
		super.onPostCreate(savedInstanceState)
		drawerToggle.syncState()
	}

	/**
	 * Listens for a tap on an item in the navigation drawer and invokes the
	 * corresponding Fragment. Also changes the title approriately.
	 */
	inner class DrawerItemClickListener(
		private val drawerLayout: DrawerLayout) : NavigationView.OnNavigationItemSelectedListener {

		override fun onNavigationItemSelected(item: MenuItem): Boolean {
			Log.d(LOG_TAG, "onNavigationItemSelected: item=" + item)
			when (item.itemId) {
				R.id.menu_item_sydney_incidents ->
					navToHazards(HazardListMode.SYDNEY)
				R.id.menu_item_regional_nsw_incidents ->
					navToHazards(HazardListMode.REGIONAL)
				R.id.menu_item_favourite_cameras ->
					navToCameras(TrafficCameraListFragment.ARG_MODE_VALUE_FAVOURITES)
				R.id.menu_item_sydney_cameras -> navToCameras(
					TrafficCameraListFragment.ARG_MODE_VALUE_SYDNEY)
				R.id.menu_item_regional_nsw_cameras -> navToCameras(
					TrafficCameraListFragment.ARG_MODE_VALUE_REGIONAL)
				R.id.menu_item_travel_times_m1 -> navToTimes(Motorway.M1)
				R.id.menu_item_travel_times_m2 -> navToTimes(Motorway.M2)
				R.id.menu_item_travel_times_m4 -> navToTimes(Motorway.M4)
				R.id.menu_item_travel_times_m7 -> navToTimes(Motorway.M7)
			} // when

			drawerLayout.closeDrawer(Gravity.START)
			return true
		} // onNavigationItemSelected()

		private fun navToHazards(mode: HazardListMode) {
			val fragment = HazardListFragment.create(mode)
			fragmentManager.beginTransaction().replace(R.id.content_main, fragment).commit()
		}

		private fun navToCameras(argCameraModeValue: Int) {
			val fragment = TrafficCameraListFragment.create(argCameraModeValue)
			fragmentManager.beginTransaction().replace(R.id.content_main, fragment).commit()
		}

		private fun navToTimes(motorway: Motorway) {
			val fragment = TravelTimesFragment.create(motorway)
			fragmentManager.beginTransaction().replace(R.id.content_main, fragment).commit()
		}
	} // class DrawItemClickListener

	companion object {
		private val CONFIG_PROPERTIES_FILE_NAME = "config.properties"
		private val LOG_TAG = MainActivity::class.java.simpleName
	}
}
